### PR TITLE
Business feature

### LINKED ISSUE #32: API Route - Create a business
# Creates a business
## Creates and returns a new business

- [x] An authenticated user is required for a successful response
- [ ] New business exists in the database after request
- [x] Business data returned includes the following:
  - id
  - owner_id
  - name
  - logo_url
  - street_address
  - city
  - region
  - postcode
  - country
  - website_url
  - phone
  - timezone
  - currency
  - about_location
- [ ] Business data returned includes business_payments associated with the business
- [ ] Error response with status 400 is given when body validations are violated

### LINKED ISSUE #33: API Route - Read a business by id
# Reads a Business
## Returns details of a business specified by its id

- [x] Successful response includes data only for the specified business
- [ ] Business data returned includes 
  - id
  - owner_id
  - name
  - logo
  - street_address
  - city
  - region
  - postcode
  - country
  - website
  - phone
  - timezone
  - currency
  - about_location
  - payment_types
- [ ] Error response with status 404 is given when a business does not exist with the provided id

### LINKED ISSUE #34: API Route - Update business by id
# Updates a Business
## Updates and returns an existing business

- [x] An authenticated user is required for a successful response
- [ ] Only the owner of the business is authorized to edit
- [ ] Business record is updated in the database after request
- [x] Business data returned includes
  - id
  - name
  - logo
  - street_address
  - city
  - region
  - postcode
  - country
  - website
  - phone
  - timezone
  - currency
  - about_location
  - payment_types
- [ ] Error response with status 400 is given when body validations are violated
- [ ] Error response with status 404 is given when a group does not exist with the provided id

### LINKED ISSUE #35: API route - Delete business by id
# Deletes a business
## Deletes an existing business

- [ ] An authenticated user is required for a successful response
- [ ] Only the owner of the business is authorized to delete
- [ ] Business record is removed from the database after request
- [ ] Success response includes a message indicating a successful deletion
- [ ] Error response with status 404 is given when a business does not exist with the provided id

### PR DESCRIPTION
**Pull request by hazeluwuz**

### CODE REVIEW

File: app/api/business_routes.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+business_routes = Blueprint('business', __name__)
+
+@business_routes.route("/<int:id>")
+def get_business_by_id(businessId):
+    """
+    Query for a business by id and return it as a dictionary
+    """
+    business = Business.query.get(businessId)
+
```
Comment (E-F-III) [inline app/api/business_routes.py:15]: Error response with status 404 is given when a business does not exist with the provided id

File: app/api/business_routes.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            telephone=form.data.telephone,
+            timezone=form.data.timezone,
+            currency=form.data.currency,
+            about_location=form.data.about_location,
+            owner_id=current_user.id
+        )
+        db.session.add(new_business)
+        db.session.commit()
+        return new_business.to_dict(), 201
+    return {'errors': validation_errors_to_error_messages(form.errors)}, 401
```
Comment (E-F-III) [inline app/api/business_routes.py:54]: StatusCode 401 = Unauthroized

400 for bad request

File: app/api/business_routes.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        business_to_edit.telephone = form.data['telephone']
+        business_to_edit.timezone = form.data['timezone']
+        business_to_edit.currency = form.data['currency']
+        business_to_edit.about_location = form.data['about_location']
+
+        db.session.add(new_business)
+        db.session.commit()
+        return new_business.to_dict(), 201
+
+    return {'errors': validation_errors_to_error_messages(form.errors)}, 401
```
Comment (E-F-III) [inline app/api/business_routes.py]: StatusCode 401 = Unauthroized

400 for bad request

File: app/api/business_routes.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+@business_routes.route("/", methods=["PUT"])
+@login_required
+def edit_business(businessId):
+    """
+    Function to edit an existing business (get business by their id)
+    """
+    form = BusinessForm()
+    form["csrf_token"].data = request.cookies["csrf_token"]
+
+    business_to_edit = Business.query.get(businessId)
```
Comment (E-F-III) [inline app/api/business_routes.py:66]: Check if business exists - 404

Check if current user is the owner of the business before editing - 401

File: app/api/business_routes.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+@business_routes.route("/", methods=["DELETE"])
+@login_required
+def delete_a_business(businessId):
+    """
+    Function to delete an existing business (get business by their id)
+    """
+    business_to_delete = Business.query.get(businessId)
+
```
Comment (E-F-III) [inline app/api/business_routes.py:110]: Check if business exists - 404

**Review by E-F-III — CHANGES_REQUESTED** [review]

**Comment by rudyn2010** [comment]
for review - ed

File: app/models/businesses.py [hunk 6]
```diff
@@ -19,6 +19,7 @@ class Business(db.Model):
     timezone = db.Column(db.String(255), nullable=False)
     currency = db.Column(db.String(255), nullable=False)
     about_location = db.Column(db.String(255), nullable=False)
+    payment_types = db.Column(db.String(255), nullable=False)
```
Comment (E-F-III) [inline app/models/businesses.py]: payment types not a column on business table. will come as associated data

File: app/api/business_routes.py (hunk 2 above)
Comment (E-F-III) [reply app/api/business_routes.py:54]: error 400 for body validation errors!

**Review by E-F-III — CHANGES_REQUESTED** [review]

**Review by E-F-III — APPROVED** [review]

**Pull request merged**